python/src/iceberg/transforms.py

```diff
@@ -0,0 +1,471 @@
+        elif self._type == FloatType or self._type == DoubleType:
+            # bucketing by Float/Double is not allowed by the spec, but they have hash implementation
+            self._hash_func = lambda v: mmh3.hash(struct.pack("d", v))
+        elif self._type == UUIDType:
+            self._hash_func = lambda v: mmh3.hash(
+                struct.pack(
+                    ">QQ",
+                    (v.int >> 64) & 0xFFFFFFFFFFFFFFFF,
+                    v.int & 0xFFFFFFFFFFFFFFFF,
+                )
+            )
+        else:
+            self._hash_func = lambda v: mmh3.hash(struct.pack("q", v))
+
+    @property
+    def num_buckets(self) -> int:
+        return self._num_buckets
+
+    def apply(self, value) -> Optional[int]:
+        if value is None:
+            return None
+
+        return (self._hash_func(value) & Bucket._MAX_32_BITS_INT) % self._num_buckets
+
+    def can_transform(self, target: Type) -> bool:
+        return self._can_transform(target)
+
+    def result_type(self, source: Type):
+        return IntegerType
+
+
+class TimeTransform(Transform):
+    """Time class is for both Date transforms and Timestamp transforms."""
+
+    _TIME_SATISFIED_ORDER = dict(year=3, month=2, day=1, hour=0)
+
+    def __init__(self, source_type: Type, name: str, apply_func: Callable[[int], int]):
+        super().__init__(
+            name,
+            f"transforms.{name}(source_type={repr(source_type)})",
+            getattr(transform_util, f"human_{name}"),
+        )
+        self._type = source_type
+        self._name = name
+        self._apply = apply_func
+        self._result_type = DateType if self._name == "day" else IntegerType
+
+    def apply(self, value: int) -> int:
+        return self._apply(value)
+
+    def can_transform(self, target: Type) -> bool:
+        if self._type == DateType:
+            return target == DateType
+        else:  # self._type is either TimestampType or TimestamptzType
+            return target == TimestampType or target == TimestamptzType
+
+    def result_type(self, source_type: Type) -> Type:
+        return self._result_type
+
+    def preserves_order(self) -> bool:
+        return True
+
+    def satisfies_order_of(self, other: Transform) -> bool:
+        if self == other:
+            return True
+
+        if isinstance(other, TimeTransform):
+            return (
+                TimeTransform._TIME_SATISFIED_ORDER[self._name]
+                <= TimeTransform._TIME_SATISFIED_ORDER[other._name]
+            )
+
+        return False
+
+    def dedup_name(self) -> str:
+        return "time"
+
+
+class Identity(Transform):
+    def __init__(
+        self,
+        source_type: Type,
+        human_str: Callable[[Any], str],
+    ):
+        super().__init__(
+            "identity",
+            f"transforms.identity(source_type={repr(source_type)})",
+            human_str,
+        )
+        self._type = source_type
```

Comment:
Why is this needed? It isn't used anywhere in this class.

Reply:
SG, removed.